rfcs/0120-no-nested-with.md

```diff
@@ -0,0 +1,94 @@
+---
+feature: no-nested-with
+start-date: 2021-01-11
+author: Alain Zscheile (@zseri)
+co-authors: @sternenseemann
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Disallow or discourage usage of multiple `with` expressions covering
+the same expression / forbid nesting `with` expressions, even indirectly.
+If infeasible in general (e.g. forbidding it in all nix expressions),
+this can be limited to nixpkgs.
+
+# Motivation
+[motivation]: #motivation
+
+It makes static analysis of nixpkgs easier, because as soon as `with`
+expressions are nested, it becomes basically impossible to [statically
+deduce where any free variable comes from] without implementing a
+full-blown nix evaluator including lazy evaluation, which is difficult as
+soon as `with` expressions and mutually recursive imports are involved
+(e.g. as currently present in `nixpkgs/lib/systems/{inspect,parse}.nix`).
+
+With this approach, any reference to any free variable can be easily
+resolved to the enclosing `with` expression "scope-include", and
+because this `with` expression couldn't then be enclosed by another
+one, even indirectly, no lookup ambiguity exists.
+
+# Detailed design
+[design]: #detailed-design
+
+At least warn about any nested usage of `with` expressions, at least when they get evaluated,
+possibly even when they get parsed. After a grace period, abort instead.
```

Comment:
This currently partially contradicts the summary and is shorter than the summary. Needs to be fixed.